storage/client/internal/parse_rfc3339.cc

```diff
@@ -0,0 +1,181 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/internal/parse_rfc3339.h"
+#include "google/cloud/internal/throw_delegate.h"
+#include <cstdio>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+namespace {
+[[noreturn]] void ReportError(std::string const& timestamp, char const* msg) {
+  std::ostringstream os;
+  os << "Error parsing RFC 3339 timestamp: " << msg
+     << " Valid format is YYYY-MM-DD[Tt]HH:MM:SS[.s+](Z|+HH:MM), got="
+     << timestamp;
+  google::cloud::internal::RaiseInvalidArgument(os.str());
+}
+
+std::chrono::system_clock::time_point ParseDateTime(
+    char const*& buffer, std::string const& timestamp) {
+  // Use std::mktime to compute the number of seconds because RFC 3339 requires
+  // working with civil time, including the annoying leap seconds, and mktime
+  // does.
+  int year, mon, mday;
+  char date_time_separator;
+  int hh, mm, ss;
+
+  int pos;
+  auto count = std::sscanf(buffer, "%4d-%2d-%2d%c%2d:%2d:%2d%n", &year, &mon,
+                           &mday, &date_time_separator, &hh, &mm, &ss, &pos);
+  // All the fields up to this point have fixed width, so total width must be:
+  constexpr int EXPECTED_WIDTH = 19;
+  constexpr int EXPECTED_FIELDS = 7;
+  if (count != EXPECTED_FIELDS or pos != EXPECTED_WIDTH) {
+    ReportError(timestamp,
+                "Invalid format for RFC 3339 timestamp detected while parsing"
+                " the base date and time portion.");
+  }
+  if (date_time_separator != 'T' and date_time_separator != 't') {
+    ReportError(timestamp, "Invalid date-time separator, expected 'T' or 't'.");
+  }
+  if (mon <= 0 or mon > 12) {
+    ReportError(timestamp, "Out of range month.");
+  }
+  if (mday <= 0 or mday > 31) {
+    ReportError(timestamp, "Out of range month day.");
```

Comment:
Do you need to check that this is a valid numeric day, given the month (and considering leap years)?

Reply:
Yes.  Fixed.